### PR TITLE
fix: make pages consistent when using sort on non-unique attributes 

### LINKED ISSUE #631: When using sort on non-unique attributes pages are not consistent
SQL does not guarantee consistent ordering when values are not unique (i.e. first names). So currently when you request the first page and then the second page it could be that you will not get back some rows because ORDER BY/LIMIT/OFFSET thought it was a good idea to return the same row in both pages missing some other row(s) with the same value for the sort property. 

This could easily be solved by always appending `id` after the last `ORDER BY` entry. This should not alter the sort order, but will guarantee that the same sort order is used every time giving back consistent pages.

### PR DESCRIPTION
**Pull request by nickschot**
Fixes #631

### CODE REVIEW

**Comment by nickschot** [comment]
Depending on how Lux handles it, there might still be a possibility for an SQL injection on the sort order, as we now have to use `orderByRaw` instead of `orderBy` (for which Knex seems to handle SQL injections). Would like your thoughts on that @zacharygolba .

**Comment by nickschot** [comment]
We might also have to do some checking. If a user orders by the PK the order by will look something like `ORDER BY pk, pk ASC`.

**Comment by zacharygolba** [comment]
Maybe we can use the [uniq util](https://github.com/postlight/lux/blob/master/src/utils/uniq.js) for this?

```javascript
const { model: { tableName, primaryKey } } = this;

this.snapshots = this.snapshots
  .filter(([method]) => method !== 'orderBy')
  .concat([
    ['orderByRaw', uniq([columnName, primaryKey])
      .map(key => `"${tableName}".${key}`)
      .join(', ') + ` ${direction}`
    ]
  ]);
```

**Comment by zacharygolba** [comment]
> there might still be a possibility for an SQL injection

As you know, the controller params are pretty locked down right now. You can only sort against a valid column name at the controller level.

**Comment by nickschot** [comment]
I've updated the tests, the two remaining tests are showing some odd behaviour. Might still be some error in the implementation. Seems to create a query with an extra orderByRaw with the PK.

File: src/packages/database/query/index.js [hunk 1]
```diff
@@ -164,10 +165,10 @@ class Query<+T: any> extends Promise {
         this.snapshots = this.snapshots
           .filter(([method]) => method !== 'orderBy')
           .concat([
-            ['orderBy', [
-              `${this.model.tableName}.${columnName}`,
-              direction
-            ]]
+            ['orderByRaw', uniq([columnName, this.model.primaryKey])
+              .map(key => `${this.model.tableName}.${key}`)
+              .join(', ') + ` ${direction}`
```
Comment (zacharygolba) [inline src/packages/database/query/index.js]: Looks like you need to add an `// eslint-disable-line prefer-template` comment to `L170`.

**Comment by zacharygolba** [comment]
> My thoughts are replacing orderBy with orderByRaw in the last() and first() functions in packages/database/query/index.js.

Good call on this. I forgot we used `orderBy` in those functions.

File: src/packages/database/query/index.js [hunk 2]
```diff
@@ -164,10 +165,11 @@ class Query<+T: any> extends Promise {
         this.snapshots = this.snapshots
           .filter(([method]) => method !== 'orderBy')
```
Comment (zacharygolba) [inline src/packages/database/query/index.js]: Just noticed this line has to change to `.filter(([method]) => method !== 'orderByRaw')`.

Comment (zacharygolba) [inline src/packages/database/query/index.js]: I just noticed this line has to change to `.filter(([method]) => method !== 'orderByRaw')`

Comment (zacharygolba) [inline src/packages/database/query/index.js]: Also [L#408](https://github.com/nickschot/lux/blob/7acf9de7699fb17e2a5f4da95d184202ee44e704/src/packages/database/query/index.js#L408) should be `return 'orderByRaw';`.

Comment (nickschot) [inline src/packages/database/query/index.js]: I was doubting about those, but forgot about them. Good call!

**Comment by zacharygolba** [comment]
👏 Great work!

**Pull request merged**